gateway/src/shard/processor/impl.rs

```diff
@@ -547,17 +542,39 @@ impl ShardProcessor {
 
                     emit::bytes(self.listeners.clone(), text.as_bytes()).await;
 
-                    break match crate::json_from_str(&mut text)
-                        .map_err(|source| Error::PayloadSerialization { source })
-                    {
-                        Ok(ser) => Ok(ser),
-                        Err(err) => {
-                            debug!("Broken JSON: {:?}", &text);
-                            Err(err)
-                        }
-                    };
+                    break Self::parse_gateway_event(&mut text);
                 }
             }
         }
     }
+
+    #[cfg(all(feature = "serde_json", not(feature = "simd-json")))]
+    fn parse_gateway_event(json: &mut str) -> Result<GatewayEvent> {
+        use serde::de::DeserializeSeed;
+        use serde_json::Deserializer;
+        use twilight_model::gateway::event::GatewayEventDeserializer;
+
+        let gateway_deserializer = GatewayEventDeserializer::from_json(json).unwrap();
+        let mut json_deserializer = Deserializer::from_str(json);
+
+        gateway_deserializer
+            .deserialize(&mut json_deserializer)
+            .map_err(|source| Error::PayloadSerialization { source })
+    }
+
+    #[allow(unsafe_code)]
+    #[cfg(feature = "simd-json")]
+    fn parse_gateway_event(json: &mut str) -> Result<GatewayEvent> {
+        use serde::de::DeserializeSeed;
+        use simd_json::Deserializer;
+        use twilight_model::gateway::event::gateway::GatewayEventDeserializerOwned;
+
+        let gateway_deserializer = GatewayEventDeserializerOwned::from_json(json).unwrap();
+        let mut json_deserializer =
+            Deserializer::from_slice(unsafe { json.as_bytes_mut() }).unwrap();
```

Comment:
Most other crates follow a convention with annotating unsafe blocks with `// Safety: Reason why this operation is safe`. See [DashMap](https://github.com/xacrimon/dashmap/blob/acrimon/experimental/src/alloc.rs#L19) for an example. 

I think the safety behind this is trivial, even the std docs don't mention any invariants that must be upheld, but referencing that could be good.